File: src/game/player/Player.js
import { BodySystem } from './BodyComponent.js';

export class Player {
    constructor(options = {}) {
        this.name = options.name || 'Player';
        this.district = options.district || null;
        this.origin = options.origin || null;
        this.generation = options.generation || 1;

        this.age = 0;
        this.lifeStage = 'childhood';

        this.body = new BodySystem();

        this.stats = {
            hp: 50,
            maxHp: 50,
            stamina: 30,
            maxStamina: 30,
            actionPoints: 2,
            perception: 10,
            fortune: 10,
            resilience: 5
        };

        this.inventory = [];
        this.money = options.startingMoney || 100;
        this.skills = [];
        this.knownRecipes = [];
        this.resonanceShards = 0;

        this.position = { x: 0, y: 0 };
        this.facing = 'down';

        this.currentDistrict = options.district;
        this.quests = [];
        this.flags = {
            birthLottery: options.birthLottery || null,
            rooted: false,
            weightGain: 0
        };
    }

    getStat(key) {
        return this.stats[key] || 0;
    }

    setStat(key, value) {
        this.stats[key] = value;
        if (key === 'maxHp') {
            this.stats.hp = Math.min(this.stats.hp, value);
        }
        if (key === 'maxStamina') {
            this.stats.stamina = Math.min(this.stats.stamina, value);
        }
    }

    modifyStat(key, delta) {
        this.stats[key] = Math.max(0, (this.stats[key] || 0) + delta);
    }

    takeDamage(amount) {
        this.stats.hp = Math.max(0, this.stats.hp - amount);
        return this.stats.hp <= 0;
    }

    heal(amount) {
        this.stats.hp = Math.min(this.stats.maxHp, this.stats.hp + amount);
    }

    useStamina(amount) {
        if (this.stats.stamina < amount) return false;
        this.stats.stamina -= amount;
        return true;
    }

    restoreStamina(amount) {
        this.stats.stamina = Math.min(this.stats.maxStamina, this.stats.stamina + amount);
    }

    addItem(item) {
        this.inventory.push(item);
    }

    removeItem(itemId) {
        const idx = this.inventory.findIndex(i => i.id === itemId);
        if (idx >= 0) {
            return this.inventory.splice(idx, 1)[0];
        }
        return null;
    }

    hasItem(itemId) {
        return this.inventory.some(i => i.id === itemId);
    }

    learnSkill(skill) {
        if (!this.skills.find(s => s.id === skill.id)) {
            this.skills.push(skill);
        }
    }

    setFlag(key, value = true) {
        this.flags[key] = value;
    }

    getFlag(key) {
        return this.flags[key] || false;
    }

    serialize() {
        return {
            name: this.name,
            district: this.district,
            origin: this.origin,
            generation: this.generation,
            age: this.age,
            lifeStage: this.lifeStage,
            body: this.body.serialize(),
            stats: { ...this.stats },
            inventory: this.inventory,
            money: this.money,
            skills: this.skills,
            knownRecipes: this.knownRecipes,
            resonanceShards: this.resonanceShards,
            position: this.position,
            facing: this.facing,
            currentDistrict: this.currentDistrict,
            quests: this.quests,
            flags: this.flags
        };
    }

    static deserialize(data) {
        const player = new Player({
            name: data.name,
            district: data.district,
            origin: data.origin,
            generation: data.generation
        });
        player.age = data.age;
        player.lifeStage = data.lifeStage;
        player.body = BodySystem.deserialize(data.body);
        player.stats = { ...data.stats };
        player.inventory = data.inventory || [];
        player.money = data.money;
        player.skills = data.skills || [];
        player.knownRecipes = data.knownRecipes || [];
        player.resonanceShards = data.resonanceShards || 0;
        player.position = data.position || { x: 0, y: 0 };
        player.facing = data.facing || 'down';
        player.currentDistrict = data.currentDistrict;
        player.quests = data.quests || [];
        player.flags = data.flags || {};
        return player;
    }
}

File: unity/Assets/Scripts/Game/Player.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace KunchengRPG.Game
{
    /// <summary>
    /// Six-component body system: left_leg, right_leg, left_hand, right_hand, torso, brain.
    /// Each component has efficiency/stability/growth, can hold equipment, and can have anomalies.
    /// </summary>
    [Serializable]
    public class BodyComponent
    {
        public string partName;
        public int efficiency;      // How strong (feeds attributes)
        public int stability;       // How stable (controls anomaly trigger rate)
        public float growth;        // Current growth progress
        public float growthMultiplier; // Per-component growth speed modifier
        public bool injured;
        public string anomaly;      // Active anomaly ID (null if clean)
        public string equipmentId;  // Equipped item ID (null if empty)
        public int equipmentDurability; // Remaining durability of equipmentId

        public bool HasAnomaly => !string.IsNullOrEmpty(anomaly);
        public bool HasEquipment => !string.IsNullOrEmpty(equipmentId);

        // Soft cap 50, hard cap 80
        public const int SOFT_CAP = 50;
        public const int HARD_CAP = 80;

        public void AddEfficiency(int amount)
        {
            efficiency = Math.Min(HARD_CAP, efficiency + amount);
        }

        public void AddStability(int amount)
        {
            stability = Math.Min(HARD_CAP, stability + amount);
        }

        public float AnomalyTriggerChance()
        {
            // stability 10 = 20%, stability 50 = 0%
            if (stability >= 50) return 0f;
            return (50 - stability) * 0.004f; // 0.4% per point below 50
        }
    }

    /// <summary>
    /// Player stats — Tier 2 (base attributes) and Tier 3 (derived attributes).
    /// </summary>
    [Serializable]
    public class PlayerStats
    {
        // Tier 2: Base attributes
        public int perception;   // 感知
        public int fortune;      // 机缘
        public int resilience;   // 韧性
        public int strength;     // 力量 (bridge value = avg of both hands)
        public int actionPower;  // 行动力
        public int vitality;     // 生命 (feeds HP)

        // Tier 3: Derived attributes (calculated)
        // Formulas live in StatFormulas so unmodified and modifier-adjusted stats
        // derive identically. These are the values before equipment/district modifiers.
        public int Attack => StatFormulas.Attack(strength);
        public float CritRate => StatFormulas.CritRate(fortune);
        public float CritDamage => StatFormulas.CritDamage(fortune);
        public float DodgeRate => StatFormulas.DodgeRate(actionPower);
        public float HitRate => StatFormulas.HitRate(perception);
        public float DamageReduction => StatFormulas.DamageReduction(resilience);
        public float AnomalyTriggerRate => StatFormulas.AnomalyTriggerRate(fortune);
    }

    /// <summary>
    /// Main player state. Persists across scenes, passed between generations via inheritance.
    /// </summary>
    [Serializable]
    public class Player
    {
        // Identity
        public string name;
        public int generation;
        public int age;
        public string districtId;
        public string birthLottery; // "native" or "drifter"

        // Vitals
        public int hp;
        public int maxHp;
        public int stamina;
        public int maxStamina;
        public int actionPoints;     // Remaining AP this year
        public int maxActionPoints;  // Total AP per year
        public int money;
        public int weight;           // Body weight (affects gameplay)

        // Systems
        public PlayerStats stats;
        public BodyComponent[] bodyComponents; // 6 components
        public List<string> inventory;         // Item IDs
        public Dictionary<string, int> flags;   // Custom flags (weightGain, etc.)
        public Dictionary<string, int> affinity; // Per-district affinity

        // Inheritance
        public string originId;
        public string propertyDistrict;
        public int resonanceShards;
        public List<Data.SkillRecipe> knownRecipes = new List<Data.SkillRecipe>();
        public List<string> discoveredPOIs = new List<string>();
        public Dictionary<string, int> familyRep = new Dictionary<string, int>();

        // Position
        public int tileX;
        public int tileY;
        public int facing; // 0=down, 1=left, 2=right, 3=up

        // Life stage
        public string LifeStage
        {
            get
            {
                if (age <= 5) return "infant";
                if (age <= 12) return "childhood";
                if (age <= 18) return "teen";
                if (age <= 35) return "young_adult";
                if (age <= 65) return "prime";
                if (age <= 80) return "middle_age";
                return "elderly";
            }
        }

        public bool IsAdult => age >= 19;

        // Actions per year based on life stage
        public int ActionsPerYear
        {
            get
            {
                switch (LifeStage)
                {
                    case "infant": return 0;
                    case "childhood": return 4;
                    case "teen": return 8;
                    default: return 14;
                }
            }
        }

        public bool HasChildren()
        {
            return flags != null && flags.ContainsKey("has_children") && flags["has_children"] > 0;
        }

        /// <summary>
        /// True when the player actively declined the marriage/family route rather than
        /// failing to reach it. This is what separates the "Free Bird" ending from
        /// "The Last Native" - both die childless, but only one of them chose to.
        /// </summary>
        public bool ChoseCelibacy()
        {
            return flags != null && flags.ContainsKey("chose_celibacy") && flags["chose_celibacy"] > 0;
        }

        /// <summary>
        /// Record the player's answer to the marriage beat. Declining is sticky: once you
        /// have said no on purpose, a later childless death still reads as a choice.
        /// Accepting clears the flag so someone who changes their mind and then fails to
        /// have a child gets "The Last Native" instead.
        /// </summary>
        public void SetCelibacyChoice(bool declined)
        {
            if (flags == null) flags = new Dictionary<string, int>();
            flags["chose_celibacy"] = declined ? 1 : 0;
        }

        public bool HasAnomaly()
        {
            if (bodyComponents == null) return false;
            foreach (var comp in bodyComponents)
            {
                if (comp != null && comp.HasAnomaly) return true;
            }
            return false;
        }

        public int GetTotalEfficiency()
        {
            int total = 0;
            foreach (var comp in bodyComponents)
                if (comp != null) total += comp.efficiency;
            return total;
        }

        public int GetTotalStability()
        {
            int total = 0;
            foreach (var comp in bodyComponents)
                if (comp != null) total += comp.stability;
            return total;
        }

        public void Heal(int amount)
        {
            hp = Math.Min(maxHp, hp + amount);
        }

        public void TakeDamage(int amount)
        {
            hp = Math.Max(0, hp - amount);
        }

        public void RestoreStamina(int amount)
        {
            stamina = Math.Min(maxStamina, stamina + amount);
        }

        public void SpendStamina(int amount)
        {
            stamina = Math.Max(0, stamina - amount);
        }

        public void AddMoney(int amount)
        {
            money = Math.Max(0, money + amount);
        }

        public bool SpendMoney(int amount)
        {
            if (money < amount) return false;
            money -= amount;
            return true;
        }

        public void ResetActionPoints()
        {
            maxActionPoints = ActionsPerYear;
            actionPoints = maxActionPoints;
        }

        public void ConsumeActionPoint()
        {
            actionPoints = Math.Max(0, actionPoints - 1);
        }

        public bool HasActionsLeft()
        {
            return actionPoints > 0;
        }
    }
}
